core/include/scipp/core/element/util.h

```diff
@@ -111,4 +111,12 @@ constexpr auto fill =
     overloaded{arg_list<double, float, std::tuple<float, double>>,
                [](auto &x, const auto &value) { x = value; }};
 
+template <typename OP> struct assign_op : OP {
+  template <typename OUT, typename... IN>
+  void operator()(OUT &out, IN &&... in) {
+    out = OP::operator()(std::forward<IN>(in)...);
+  }
+};
+template <typename OP> assign_op(OP) -> assign_op<OP>;
```

Comment:
Moving `assign_op` into `transform_common.h` would move it out of the `element` namespace. Do we want that?

Reply:
Not sure. It could actually just be made part of `scipp/common/` since it is very generic? On the other hand, we may eventually want to add support for `transform_flags` and `arg_list`. Maybe just a completely new header? Or merge (and rename) `element/arg_list.h`?

Reply:
I've had ideas for 'transform adapters' which could e.g. chain kernels. This would allow us to implement kernel fusion which might, once the rest is optimised enough, improve performance. But this is probably a long way into the future. But `assign_op` seems to fall into the same category.

Reply:
I feel the compiler would generally deal well with chained kernels well enough, without need for any trickery. The main performance hurdle right now is the multi-dimensional indexing. See #1363:

> Performance in transform. We may need to add special case handling in transform, e.g., for stride 1, so the compiler can generate better code. It is currently unclear whether this is an issue in practice. There is a really significant slowdown in the single-threaded case, but we may still be good enough when using threads to be memory-bandwidth limited.

We will likely also want stride 0 optimization, which is highly relevant for all event-data arithmetic, for example in unit conversions.

It should in principle not be too hard to do this, but we may need to take steps to avoid too much code generation for cases with many arguments. Also, it is not clear that this is the only/correct approach. Maybe we just need a better `MultiIndex`? Better compiler flags for better inlining? Hand written nested loops?

It is probably a longer discussion/analysis to develop a complete understanding of all the cases and requirements, and weight cost/benefit of the various solutions.

Reply:
I agree. I just wanted to put it out there in case it helps us find a good place to put `assign_op`.

Unfortunately, I don't see any file it really fits into at the moment. We could just put it into `transform_common.h` for now.

Reply:
> We could just put it into transform_common.h for now.

👍